google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java

```diff
@@ -1299,6 +1304,24 @@ public void packageGetterSerialize() {
     assertJson("{'publicValue': 'bar'}", serialize(bean));
   }
 
+  @Test
+  public void serializeOptional() {
+    assertEquals("foo", serialize(Optional.of("foo")));
+  }
+
+  @Test
+  public void deserializeOptional() {
+    OptionalStringBean bean = deserialize("{'foo': 'bar'}", OptionalStringBean.class);
```

Comment:
Hi and thanks a lot for taking a look at the code. You raise an interesting point which I overlooked at the first Iteration!

When receiving null in our JSON, I think it best makes sense to set the property to Optional.empty() and not to null. Also, there's no distinction between the first and the second case you described, as Optional.empty() and Optional.ofNullable(null) are equivalent.

I've adjusted my implementation and added a test case for the described behaviour.

Reply:
Thank you very much for your contribution. We are discussing the best way to handle some edge cases right now. 
Just want to check with you that you have any use case for serializing something like `Optional<Optional<String>>`?
Or you think we should just not support the nested Optional object's serialization:

```
  private static class OptionalInnerdBean{
    public Optional<String> bar;
  }
  private static class OptionalNestBean{
    public Optional<OptionalInnerdBean> foo;
  }

  @Test
  public void deserializeNestedOptional() {
    OptionalNestBean outer = new OptionalNestBean();
    OptionalInnerdBean inner = new OptionalInnerdBean();
    inner.bar = Optional.of("foo-bar");
    outer.foo = Optional.of(inner);
    
    assertExceptionContains(
            "throw some error message",
            () -> serialize(outer));
  }
  ```

Reply:
I guess the best way to serialize null to `Optional<Optional<String>>` would be `Optional.of(Optional.empty())`. Would like to have any other modifications or other edge cases handled? Then I'll adjust the code accordingly and also add a test (incl. implementation) for the case you just described.

Reply:
Hi, thank you very much for the update. Apologize for getting back to you late. 

We had some meetings within the team, and decided we are going to support serialization of `Optional<T>` in a different way. There are a few reasons why we hesitate about direct supporting serialization of `Optional<T>`:

1. We need to keep our Java server SDK aligned with Android SDK (but this reflection requires android API level 29, which is too high for now );
2. Different customers might want to handle the serialization of `Optional` in different approach, we are leaning to provide a mechanism to customers so that they can build their own custom serializers in the future to handle these cases.

Reply:
Okay, thanks a lot for the feedback! I guess I can close this PR now.